backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/CacheSupport.kt

```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright 2010-2019 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license
+ * that can be found in the LICENSE file.
+ */
+
+package org.jetbrains.kotlin.backend.konan
+
+import org.jetbrains.kotlin.config.CompilerConfiguration
+import org.jetbrains.kotlin.konan.library.KonanLibrary
+import org.jetbrains.kotlin.konan.library.resolver.KonanLibraryResolveResult
+import org.jetbrains.kotlin.konan.target.CompilerOutputKind
+import org.jetbrains.kotlin.konan.target.KonanTarget
+
+class CacheSupport(
+        configuration: CompilerConfiguration,
+        resolvedLibraries: KonanLibraryResolveResult,
+        target: KonanTarget,
+        produce: CompilerOutputKind
+) {
+
+    internal val cachedLibraries: CachedLibraries = run {
+        val allLibraries = resolvedLibraries.getFullList()
+
+        val explicitCacheFiles = configuration.get(KonanConfigKeys.CACHED_LIBRARIES)!!
+
+        val explicitCaches = allLibraries.mapNotNull { library ->
+            val cache = explicitCacheFiles[library.libraryFile.absolutePath]
+            cache?.let { library to it }
+        }.toMap()
+
+        CachedLibraries(
+                target = target,
+                allLibraries = allLibraries,
+                explicitCaches = explicitCaches,
+                implicitCacheDirectories = configuration.get(KonanConfigKeys.CACHE_DIRECTORIES)!!
+        )
+    }
+
+    internal val librariesToCache: Set<KonanLibrary> = run {
+        val files = configuration.get(KonanConfigKeys.LIBRARIES_TO_CACHE)!!.toSet()
+
+        // TODO: consider using [FeaturedLibraries.kt].
+        resolvedLibraries.getFullList()
+                .filter { it.libraryFile.absolutePath in files }
+                .toSet()
+                .also { if (!produce.isCache) check(it.isEmpty()) }
```

Comment:
Message can be more helpful than "Check failed."

Reply:
The user-friendly check for this is in `K2Native.kt`.
See also https://github.com/JetBrains/kotlin-native/pull/3359#discussion_r331409348